src/util/listQueryBuilder.js

```diff
@@ -451,6 +447,119 @@ async function performAggregationQuery ({
   return paginationMeta
 }
 
+/**
+ * Generic query function to address filters, sort and pagination for stats endpoints
+ * @param {Object}  params.queryBuilder - Knex.js query builder
+ * @param {String}  params.schema
+ * @param {String}  params.table
+ * @param {String}  [params.timeFilter = 'createdDate'] - Other filters will be disabled
+ *    if date filter is beyond retention period
+ * @param {String}  [params.timeColumn = 'createdTimestamp'] - Internal name of time column
+ * @param {String}  params.retentionLimitDate - Beyond that limit, all filters (except time) will be disabled
+ * @param {String}  params.groupBy - Field to group by (must be 'hour', 'day' or 'month')
+ * @param {Object}  [params.filters] - Filters for the query (like select on some IDs)
+ * @param {Object}  [params.filters[key]]
+ * @param {String}  [params.filters[key].dbField]
+ * @param {String}  [params.filters[key].type]
+ * @param {String|Function}  [params.filters[key].transformValue]
+ * @param {String|Function}  [params.filters[key].query]
+ *
+ * @param {Object}  params.paginationConfig
+ * @param {Number}  params.paginationConfig.page
+ * @param {Number}  params.paginationConfig.nbResultsPerPage
+ *
+ * @param {Object}  params.orderConfig
+ * @param {Object}  params.orderConfig.order - Must be 'asc' or 'desc'
+ *
+ * @return {Object} paginationMeta
+ * @return {Number} paginationMeta.nbResults
+ * @return {Number} paginationMeta.nbPages
+ * @return {Number} paginationMeta.page
+ * @return {Number} paginationMeta.nbResultsPerPage
+ * @return {Object[]} paginationMeta.results
+ */
+async function performHistoryQuery ({
+  queryBuilder,
+  schema,
+  table,
+  timeFilter = 'createdDate',
+  timeColumn = 'createdTimestamp',
+  retentionLimitDate,
+  groupBy,
+  filters,
+  paginationConfig,
+  orderConfig
+}) {
+  checkOrderConfig(orderConfig)
+
+  let onlyTimeFilter = true
+
+  if (filters) {
+    checkFilters(filters)
+
+    const nonTimeFilterKeys = _.without(Object.keys(filters), [timeFilter])
+    onlyTimeFilter = nonTimeFilterKeys.map(key => _.get(filters, `${key}.value`)).every(_.isUndefined)
+    const minRangeValue = getMinRangeValue(filters[timeFilter].value)
+
+    if (minRangeValue && minRangeValue < retentionLimitDate && !onlyTimeFilter) {
+      throw createError(400, `All filters are disabled before ${retentionLimitDate}`)
+    }
+  }
+  checkPaginationConfig(paginationConfig)
+
+  if (!['hour', 'day', 'month'].includes(groupBy)) throw new Error(`Invalid groupBy value: ${groupBy}`)
+
+  const intervals = {
+    hour: '1 hour',
+    day: '1 day',
+
+    // TimescaleDB doesn't support month interval (https://github.com/timescale/timescaledb/issues/414)
+    month: '30 days'
```

Comment:
This is embarassing as it was on of the requirements for the migration.
Please ensure it’s specified in the updated API reference (have the draft ready before merging).